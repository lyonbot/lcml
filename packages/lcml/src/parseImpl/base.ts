import { ParseError } from '../ParseError';
import type { StringStream } from '../StringStream';

const $$isParsedNode = Symbol();

export interface ParsedNodeBase {
  [$$isParsedNode]: true;
  type: string;
  start: number;
  end: number;
  parsedLength: number;
}

export function isParsedNode(x: any): x is ParsedNodeBase {
  return typeof x === 'object' && x !== null && x[$$isParsedNode] === true;
}

const defaultCtxOpts = {
  recover: false, // if false, when met error, all remainder will be skipped
  onNode: null as ((node: ParsedNodeBase) => void) | null,
  onError: null as ((error: ParseError) => void) | null,
};
const ctxOpts = { ...defaultCtxOpts };

export function startParsing(opts: Partial<typeof ctxOpts>) {
  Object.assign(ctxOpts, defaultCtxOpts, opts);
  _isHalting = null;
}
export function endParsing() {
  Object.assign(ctxOpts, defaultCtxOpts);
  _isHalting = null;
}

// while recovering,
// array and object's parsing processes are stopped
// and partially parsed result is generated, returned

let _isHalting: ParseError | null = null;
export const isHaltingParser = () => _isHalting;

export function makeParsedNode<T extends ParsedNodeBase>(
  sourceStream: StringStream,
  extra: Omit<T, 'start' | 'end' | typeof $$isParsedNode>,
): T {
  const parsedLength = extra.parsedLength;

  const answer = {
    ...extra,
    start: sourceStream.pos,
    end: sourceStream.pos + parsedLength,
  } as T;

  Object.defineProperty(answer, $$isParsedNode, {
    configurable: true,
    enumerable: false,
    value: true,
  });

  if (ctxOpts.onNode) ctxOpts.onNode(answer);
  return answer;
}

/**
 * if current parsing process disallows recovering,
 * this set internal `_isHalting` flag to true
 */
export function commitParseError(...args: ConstructorParameters<typeof ParseError>) {
  const error = new ParseError(...args);
  if (ctxOpts.onError) ctxOpts.onError(error);
  if (!ctxOpts.recover) _isHalting = error;
}
