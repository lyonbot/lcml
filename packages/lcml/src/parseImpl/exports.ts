export { ParseError } from '../ParseError';
export { isParsedNode, ParsedNodeBase } from './base';
export * from './nodeTypes';
