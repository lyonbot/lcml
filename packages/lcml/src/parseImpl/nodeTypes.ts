import type { ParsedArrayNode } from './parseArray';
import type { ParsedCommentNode } from './parseComment';
import type { ParsedExpressionNode } from './parseExpression';
import type { ParsedBooleanNode, ParsedNullishNode, ParsedNumberNode, ParsedLiteralNode } from './parseLiteral';
import type { ParsedObjectPropertyNode, ParsedObjectNode } from './parseObject';
import type { ParsedStringNode, ParsedStringSegmentNode } from './parseString';
import type { ParsedIdentifierNode } from './parseIdentifier';
import type { ParsedValueNode } from './parseValue';

export interface ParsedNodeTypeLUT {
  'array': ParsedArrayNode;
  'comment': ParsedCommentNode;
  'expression': ParsedExpressionNode;
  'boolean': ParsedBooleanNode;
  'nullish': ParsedNullishNode;
  'number': ParsedNumberNode;
  // "literal": ParsedLiteralNode;
  'object-property': ParsedObjectPropertyNode;
  'object': ParsedObjectNode;
  'string': ParsedStringNode;
  'string-segment': ParsedStringSegmentNode;
  'identifier': ParsedIdentifierNode;
  // "value": ParsedValueNode;
}

type TempCheck1 = { [x in keyof ParsedNodeTypeLUT]: ParsedNodeTypeLUT[x] extends { type: x } ? never : x };
type BadItemsInLUT = TempCheck1[keyof TempCheck1];

// eslint-disable-next-line @typescript-eslint/no-unused-vars
const _check__noBadItemsInLUT: BadItemsInLUT extends never ? true : false = true;

export type ParsedNode = ParsedNodeTypeLUT[keyof ParsedNodeTypeLUT];

export {
  ParsedArrayNode,
  ParsedCommentNode,
  ParsedExpressionNode,
  ParsedBooleanNode,
  ParsedNullishNode,
  ParsedNumberNode,
  ParsedLiteralNode,
  ParsedObjectPropertyNode,
  ParsedObjectNode,
  ParsedStringNode,
  ParsedStringSegmentNode,
  ParsedIdentifierNode,
  ParsedValueNode,
};
