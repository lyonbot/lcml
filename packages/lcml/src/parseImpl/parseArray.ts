import { ParsedNodeBase, makeParsedNode, isRecovering, makePanic } from './base';
import { parseComment } from './parseComment';
import { StringStream } from '../StringStream';
import { parseValue, ParsedValueNode } from './parseValue';

export interface ParsedArrayNode extends ParsedNodeBase {
  type: 'array';
  length: number;
  items: (ParsedValueNode | null)[];
  itemsLocation: {
    hasValue: boolean;
    valueStart: number;
    valueEnd: number;

    hasTrailingComma?: boolean;
    trailingCommaStart?: number;
  }[];
}

export function parseArray(stream: StringStream, top: string): ParsedArrayNode | void {
  if (top !== '[') return;

  const ss = stream.clone(1);

  const items = [] as ParsedArrayNode['items'];
  const itemsLocation = [] as ParsedArrayNode['itemsLocation'];

  let index = 0;
  let slotEmpty = true;
  let finished = false;

  const finalize = () => {
    let length = itemsLocation.length;
    if (length > 2 && !itemsLocation[length - 1].hasValue) {
      // remove invalid empty slot made by trailing comma
      length--;
      itemsLocation.pop();
    }

    return makeParsedNode<ParsedArrayNode>(stream, {
      type: 'array',
      parsedLength: ss.pos - stream.pos,
      length,
      items: itemsLocation.map((_, i) => items[i] || null),
      itemsLocation,
    });
  };

  while (!isRecovering() && !finished && (ss.skipSpaces(), !ss.eof())) {
    const top = ss.peek();
    if (top === ']') {
      ss.precede(1);
      finished = true;
      break;
    }

    if (top === ',') {
      itemsLocation[index] = {
        ...(itemsLocation[index] || {
          hasValue: false,
          valueStart: ss.pos,
          valueEnd: ss.pos,
        }),
        hasTrailingComma: true,
        trailingCommaStart: ss.pos,
      };

      index++;
      slotEmpty = true;

      ss.precede(1);

      // after updating index,
      // prepare the next item's location info
      // this is useful for empty slot
      itemsLocation[index] = {
        hasValue: false,
        valueStart: ss.pos,
        valueEnd: ss.pos,
      };

      continue;
    }

    const comment = parseComment(ss, top);
    if (comment) {
      ss.precede(comment.parsedLength);
      continue;
    }

    const item = slotEmpty && parseValue(ss, top);
    if (!item) {
      return makePanic(finalize, `expect ${slotEmpty ? 'value, ' : ''}comma or right square bracket`, ss);
    }

    items[index] = item;
    itemsLocation[index] = {
      ...itemsLocation[index],
      hasValue: true,
      valueStart: ss.pos,
      valueEnd: ss.pos + item.parsedLength,
    };

    slotEmpty = false;
    ss.precede(item.parsedLength);
  }

  if (!isRecovering() && !finished) {
    return makePanic(finalize, 'expect right square bracket', ss);
  }

  return finalize();
}
