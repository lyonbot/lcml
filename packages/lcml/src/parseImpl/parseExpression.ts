import { ParsedNodeBase, makeParsedNode } from './base';
import { StringStream } from '../StringStream';

export interface ParsedExpressionNode extends ParsedNodeBase {
  type: 'expression';
  expression: string;
  expressionStart: number;
  expressionEnd: number;
  leftBracket: string;
  rightBracket: string;
}

export function parseExpression(stream: StringStream, top: string): ParsedExpressionNode | void {
  if (top !== '{') return;
  if (stream.peek(2) !== '{{') return;

  // we support {{{ expression }}}
  let bracketLength = 2;
  while (stream.substr(bracketLength, 1) === '{') bracketLength++;

  const leftBracket = stream.substr(0, bracketLength);
  const rightBracket = '}'.repeat(bracketLength);

  const endToken = stream.indexOf(rightBracket, bracketLength);
  if (!endToken) return;

  return makeParsedNode<ParsedExpressionNode>(stream, {
    type: 'expression',
    parsedLength: endToken.endOffset,
    expression: stream.substr(bracketLength, endToken.offset - bracketLength),
    expressionStart: bracketLength + stream.pos,
    expressionEnd: endToken.offset + stream.pos,
    leftBracket,
    rightBracket,
  });
}
