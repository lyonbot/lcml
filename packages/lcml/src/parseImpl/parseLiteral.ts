import { ParsedNodeBase, makeParsedNode } from './base';
import { StringStream } from '../StringStream';
import { isNonWordCharacter } from '../string';

export interface ParsedNumberNode extends ParsedNodeBase {
  type: 'number';
  raw: string;
  value: number;
}

export interface ParsedBooleanNode extends ParsedNodeBase {
  type: 'boolean';
  raw: string;
  value: boolean;
}

export interface ParsedNullishNode extends ParsedNodeBase {
  type: 'nullish';
  /** @example - `'null' | 'undefined'` */
  raw: string;
}

export type ParsedLiteralNode = ParsedNumberNode | ParsedBooleanNode | ParsedNullishNode;

export function parseLiteral(stream: StringStream, top: string): ParsedLiteralNode | void {
  const matchRaw = (s: string) =>
    top === s[0] && stream.peek(s.length) === s && isNonWordCharacter(stream.substr(s.length, 1)) ? s : false;

  let raw: string | false;

  if ((raw = matchRaw('true') || matchRaw('false'))) {
    return makeParsedNode<ParsedBooleanNode>(stream, {
      type: 'boolean',
      parsedLength: raw.length,
      raw,
      value: raw === 'true',
    });
  }

  if ((raw = matchRaw('null') || matchRaw('undefined'))) {
    return makeParsedNode<ParsedNullishNode>(stream, {
      type: 'nullish',
      parsedLength: raw.length,
      raw,
    });
  }

  if ((raw = matchRaw('Infinity') || matchRaw('NaN'))) {
    return makeParsedNode<ParsedNumberNode>(stream, {
      type: 'number',
      parsedLength: raw.length,
      raw,
      value: raw === 'NaN' ? NaN : Infinity,
    });
  }

  if ((raw = matchRaw('-Infinity') || matchRaw('+Infinity'))) {
    return makeParsedNode<ParsedNumberNode>(stream, {
      type: 'number',
      parsedLength: raw.length,
      raw,
      value: raw[0] === '-' ? -Infinity : Infinity,
    });
  }

  // ----------------------
  if (top === '+' || top === '-' || top === '.' || (top >= '0' && top <= '9')) {
    const remainder = stream.substr(0);
    const tmp =
      /^[+-]?0(?:b[01]+|o[0-7]+|x[0-9a-fA-F]+)(?![\w.])/.exec(remainder) || // hex and octal
      /^[+-]?(?:\d+\.\d*|\.\d+|\d+)(?:[eE][+-]?\d+)?(?![\w.])/.exec(remainder); // regular number

    if (tmp) {
      raw = tmp[0];
      return makeParsedNode<ParsedNumberNode>(stream, {
        type: 'number',
        parsedLength: raw.length,
        raw,
        value: parseFloat(raw),
      });
    }
  }
}
