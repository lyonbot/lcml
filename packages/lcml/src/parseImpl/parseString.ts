import { ParsedNodeBase, makeParsedNode, commitParseError } from './base';
import { parseExpression, ParsedExpressionNode } from './parseExpression';
import { StringStream } from '../StringStream';

export interface ParsedStringSegmentNode extends ParsedNodeBase {
  type: 'string-segment';
  raw: string;
}

export interface ParsedStringNode extends ParsedNodeBase {
  type: 'string';
  /**
   * this array must has length 1,3,5...
   * - odd (0,2,4) items are `string-segment` (containing raw string with possible escape sequences)
   * - even (1,3) items are `expression`
   */
  segments: (ParsedExpressionNode | ParsedStringSegmentNode)[];
  quote: string;
  isDynamic: boolean;
}

export function parseString(stream: StringStream, top: string): ParsedStringNode | void {
  if (!(top === '"' || top === "'")) return;

  return internalParseStringContent(stream, top);
}

/**
 * assuming the leading quote is consumed.
 * parse remainder as string content
 *
 * @param quote - optional. if empty string, will parse until EOF
 */
export function internalParseStringContent(stream: StringStream, quote: string): ParsedStringNode | void {
  const needles = ['\\', quote, '{'];
  const segments: ParsedStringNode['segments'] = [];

  const ss = stream.clone(quote.length);

  const finalize = () =>
    makeParsedNode<ParsedStringNode>(stream, {
      type: 'string',
      parsedLength: ss.pos - stream.pos,
      segments,
      quote,
      isDynamic: segments.length > 1,
    });

  let indexOfSince = 0;

  const pushStringSegmentAndPrecede = (len: number) => {
    segments.push(
      makeParsedNode<ParsedStringSegmentNode>(ss, {
        type: 'string-segment',
        parsedLength: len,
        raw: ss.substr(0, len),
      }),
    );

    indexOfSince = 0;
    ss.precede(len);
  };

  // eslint-disable-next-line no-constant-condition
  while (true) {
    const tk = ss.indexOf(needles, indexOfSince);

    if (!tk) {
      pushStringSegmentAndPrecede(ss.remainderLength);
      if (!quote) {
        // no quote char, normally parse to the EOF
        return finalize();
      } else {
        // expect a quote but not found, throw an error
        commitParseError('expect end of string', stream, stream.raw.length);
        break
      }
    }

    if (tk.needle === '\\') {
      // escaped char, skip its next sibling char
      indexOfSince = tk.endOffset + 1;
      continue;
    }

    if (tk.needle === quote) {
      // find the end of the string
      pushStringSegmentAndPrecede(tk.offset);
      ss.precede(quote.length); // precede quote
      break;
    }

    if (tk.needle === '{') {
      // find the beginning of expression
      const expr = parseExpression(ss.clone(tk.offset), '{');
      if (!expr) {
        indexOfSince = tk.endOffset;
        continue;
      }

      pushStringSegmentAndPrecede(tk.offset);
      segments.push(expr);
      ss.precede(expr.parsedLength);

      continue;
    }
  }

  // successfully get segments
  return finalize();
}
