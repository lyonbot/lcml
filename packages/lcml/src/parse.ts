import { parseComment } from './parseImpl/parseComment';
import { parseValue } from './parseImpl/parseValue';
import { ParseError } from './ParseError';
import { StringStream } from './StringStream';
import { endParsing, isHaltingParser, startParsing } from './parseImpl/base';
import { ParsedExpressionNode } from './parseImpl/exports';
import { internalParseStringContent } from './parseImpl/parseString';

/**
 * parser options
 *
 * @public
 */
export interface ParseOptions {
  /**
   * treat invalid inputs as string. see README
   */
  loose?: boolean;

  /**
   * when input is empty, what shall be returned?
   *
   * @default "as-undefined"
   */
  treatEmptyInput?: 'as-undefined' | 'as-empty-string';

  /**
   * after parsing something, if there is still characters left unparsed, shall we just ignore them, or treat as error
   */
  ignoreUnparsedRemainder?: boolean;

  /**
   * when error occurs
   *
   * - `"throw"` (default) - a ParseError will be thrown
   * - `"return"` - stop parsing. the ParseError will be returned along with partially-parsed ast
   * - `"recover"` - try to continue parsing. mulitple `ParseError`s will be returned along with partially-parsed ast 
   * - `"as-string"` - discard all parsed structures and treat whole input as a string with expressions
   *
   * @default "throw"
   */
  onError?: 'throw' | 'return' | 'recover' | 'as-string';
}

export type ParseResult = ReturnType<typeof parse>;

/**
 * skip leading comments and parse the value
 *
 * @public
 */
export function parse(str: string, opts: ParseOptions = {}) {
  const expressions = [] as ParsedExpressionNode[];
  const errors = [] as ParseError[];

  try {
    const addError = (error: ParseError): void => {
      errors.push(error);
    };

    startParsing({
      recover: opts.onError === 'recover',
      onNode: node => {
        if (node.type === 'expression') expressions.push(node as ParsedExpressionNode);
      },
      onError: addError,
    });

    let stream = new StringStream(str);

    // skip leading comments

    while ((stream.skipSpaces(), !stream.eof())) {
      const top = stream.peek();

      const t = parseComment(stream, top);
      if (!t) break;
      stream.precede(t.parsedLength);
    }

    // start consuming value

    const actualStart = stream.pos;

    let isInputEmpty = false;

    if (stream.eof()) {
      // after removing comments
      // nothing left
      // we shall make a fake LCML input

      const placeholder = opts.treatEmptyInput === 'as-empty-string' ? '""' : 'undefined';
      stream = new StringStream(new Array(actualStart).fill(' ').join('') + placeholder);
      stream.precede(actualStart);
      isInputEmpty = true; //later fix the location info
    }

    const top = stream.peek();
    let ast = parseValue(stream, top);

    if (ast) {
      // successfully parsed something

      if (isInputEmpty) {
        // input LCML is fake!
        ast.parsedLength = 0;
        ast.end = ast.start;
      }

      stream.precede(ast.parsedLength);
      stream.skipSpaces();
    } else {
      // nothing parsed, we shall throw a error
      addError(new ParseError('invalid input', stream));
    }

    // see README the "Loose Mode" part's Note
    const firstType = ast && ast.type;
    const mayLoose = !!opts.loose && firstType !== 'string' && firstType !== 'array' && firstType !== 'object';

    // do extra check here: is EOF reached?
    // by default, treat unparsed part as error
    // but they can also be ignored
    if (!errors.length && !stream.eof() && (mayLoose || !opts.ignoreUnparsedRemainder)) {
      // special note: when loose mode works,
      // always make a error to trigger following as-string logic
      addError(new ParseError('unexpected remainder', stream));
    }

    // now convert
    const asString = !!errors.length && (mayLoose || opts.onError === 'as-string');
    if (asString) {
      // discard all parsed structures
      // treat whole input as string
      const stream = new StringStream(str);
      stream.precede(actualStart);
      ast = internalParseStringContent(stream, '');

      if (mayLoose) errors.splice(0); // remove all errors
    }

    // finally throw error if needed
    if (errors.length && (opts.onError || 'throw') === 'throw') {
      throw errors[0]!;
    }

    return {
      ast,
      expressions,
      actualStart,
      end: actualStart + (ast ? ast.parsedLength : 0),
      error: errors[0] as ParseError | null,
      errors,
      looseModeEnabled: asString && mayLoose,
    };
  } finally {
    endParsing();
  }
}
