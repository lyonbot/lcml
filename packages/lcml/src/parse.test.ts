import { parse } from './parse';
import { ParsedNumberNode } from './parseImpl/parseLiteral';
import { ParsedArrayNode } from './parseImpl/parseArray';
import { ParsedStringNode, ParsedStringSegmentNode } from './parseImpl/parseString';
import { ParsedExpressionNode } from './parseImpl/parseExpression';

describe.only('parse', () => {
  const parseAST = (x: string) => {
    const ans = parse(x);
    if (ans.error) throw ans.error;
    return ans.ast!;
  };

  it('parse expression', () => {
    const ast = parseAST(' {{ expr }}') as ParsedExpressionNode;

    expect(ast.type).toBe('expression');
    expect(ast.expression).toBe(' expr ');
  });

  it('parse expression-leading string', () => {
    const ast = parseAST(' {{ expr }}, welcome ! // not-comment') as ParsedStringNode;

    expect(ast.type).toBe('string');
    expect(ast).toMatchSnapshot();
  });

  it('parse string', () => {
    const ast = parseAST(' /* string */   "test \\{{ test }} {{ test2 }}"') as ParsedStringNode;

    expect(ast.type).toBe('string');
    expect(ast.quote).toBe('"');
    expect(ast.segments).toHaveLength(3);
    expect(ast.isDynamic).toBe(true);

    const [s1, s2, s3] = ast.segments as [ParsedStringSegmentNode, ParsedExpressionNode, ParsedStringSegmentNode];

    expect(s1.type).toBe('string-segment');
    expect(s1.raw).toBe('test \\{{ test }} ');

    expect(s2.type).toBe('expression');
    expect(s2.expression).toBe(' test2 ');

    expect(s3.type).toBe('string-segment');
    expect(s3.raw).toBe('');

    expect(() => parseAST(`"foo`)).toThrowError('end of string');
    expect(() => parseAST(`"foo {{ bar`)).toThrowError('end of string');
  });

  it('parse static string', () => {
    const ast = parseAST(' /* string */   "test \\{{ test }} \\"s {{ tt"') as ParsedStringNode;

    expect(ast.type).toBe('string');
    expect(ast.isDynamic).toBe(false);
    expect(ast.segments.length).toBe(1);
    expect((ast.segments[0] as ParsedStringSegmentNode).raw).toBe(`test \\{{ test }} \\"s {{ tt`);

    expect(parseAST('""')).toMatchSnapshot('empty string');
    expect(parseAST("''")).toMatchSnapshot('empty string');
  });

  it('parse number', () => {
    const goodCases = ['-1.3', '1.4e+4', '.9e-2', '+155.', '0x7F2A'];
    const badCases = ['-1.3.', '1.4e', '0xfg', '3.e10.3'];

    goodCases.forEach(c => {
      let ast: ParsedNumberNode;

      ast = parseAST(c) as ParsedNumberNode;

      expect(ast.type).toBe('number');
      expect(ast.raw).toBe(c);

      ast = parseAST('/* comment */' + c) as ParsedNumberNode;

      expect(ast.type).toBe('number');
      expect(ast.raw).toBe(c);
    });

    badCases.forEach(c => {
      try {
        const ast = parseAST(c);
        expect(ast?.type).not.toBe('number');
      } catch (err) {
        expect((err as Error).message).toContain('invalid input');
      }
    });
  });

  it('parse array', () => {
    const ast = parseAST('["test", {{ expr }}, , /* test */,]') as ParsedArrayNode;

    expect(ast.type).toBe('array');
    expect(ast.length).toBe(4);
    expect(ast.itemsLocation.map(x => !!x.hasValue)).toEqual([true, true, false, false]);
    expect(ast.itemsLocation.map(x => !!x.hasTrailingComma)).toEqual([true, true, true, true]);
    expect(ast.items[0]!.type).toBe('string');
    expect(ast.items[1]!.type).toBe('expression');
    expect(ast.items[2]).toBe(null);
    expect(ast.items[3]).toBe(null);

    expect(parseAST('[]')).toMatchSnapshot('empty array');

    expect(() => parseAST(`[`)).toThrowError('right square bracket');
    expect(() => parseAST(`[1`)).toThrowError('right square bracket');
    expect(() => parseAST(`[1,`)).toThrowError('right square bracket');
    expect(() => parseAST(`[1 2`)).toThrowError('(at :3)');
  });

  it('parse object', () => {
    const ast = parseAST(`
      {
        /*test*/
        "foo" /*test*/ : "test",
        {{ sym }}: [1,2],
        "bar": true
      }
    `);

    expect(ast).toMatchSnapshot();

    expect(parseAST(`{}`)).toMatchSnapshot('empty object');

    expect(() => parseAST(`{ "foo": // test`)).toThrowError('property value');
    expect(() => parseAST(`{ "foo": 123`)).toThrowError('right curly bracket');
  });
});
