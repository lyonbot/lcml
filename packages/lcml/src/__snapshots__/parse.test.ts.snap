// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse onError: recover 1`] = `
Object {
  "end": 95,
  "parsedLength": 78,
  "properties": Array [
    Object {
      "colonStart": 31,
      "end": 46,
      "hasColon": true,
      "hasTrailingComma": true,
      "key": Object {
        "end": 31,
        "parsedLength": 5,
        "raw": "hello",
        "start": 26,
        "type": "identifier",
      },
      "parsedLength": 20,
      "start": 26,
      "trailingCommaStart": 45,
      "type": "object-property",
      "value": null,
    },
    Object {
      "colonStart": 58,
      "end": 90,
      "hasColon": true,
      "hasTrailingComma": false,
      "key": Object {
        "end": 58,
        "parsedLength": 5,
        "raw": "world",
        "start": 53,
        "type": "identifier",
      },
      "parsedLength": 37,
      "start": 53,
      "type": "object-property",
      "value": Object {
        "end": 90,
        "items": Array [
          Object {
            "end": 63,
            "parsedLength": 1,
            "raw": "1",
            "start": 62,
            "type": "number",
            "value": 1,
          },
          Object {
            "end": 66,
            "parsedLength": 1,
            "raw": "2",
            "start": 65,
            "type": "number",
            "value": 2,
          },
          null,
        ],
        "itemsLocation": Array [
          Object {
            "hasTrailingComma": true,
            "hasValue": true,
            "trailingCommaStart": 63,
            "valueEnd": 63,
            "valueStart": 62,
          },
          Object {
            "hasTrailingComma": true,
            "hasValue": true,
            "trailingCommaStart": 66,
            "valueEnd": 66,
            "valueStart": 65,
          },
          Object {
            "hasTrailingComma": true,
            "hasValue": false,
            "trailingCommaStart": 77,
            "valueEnd": 67,
            "valueStart": 67,
          },
        ],
        "length": 3,
        "parsedLength": 30,
        "start": 60,
        "type": "array",
      },
    },
  ],
  "start": 17,
  "type": "object",
}
`;

exports[`parse parse array: empty array 1`] = `
Object {
  "end": 2,
  "items": Array [],
  "itemsLocation": Array [],
  "length": 0,
  "parsedLength": 2,
  "start": 0,
  "type": "array",
}
`;

exports[`parse parse object 1`] = `
Object {
  "end": 112,
  "parsedLength": 105,
  "properties": Array [
    Object {
      "colonStart": 49,
      "end": 58,
      "hasColon": true,
      "hasTrailingComma": true,
      "key": Object {
        "end": 39,
        "isDynamic": false,
        "parsedLength": 5,
        "quote": "\\"",
        "segments": Array [
          Object {
            "end": 38,
            "parsedLength": 3,
            "raw": "foo",
            "start": 35,
            "type": "string-segment",
          },
        ],
        "start": 34,
        "type": "string",
      },
      "parsedLength": 24,
      "start": 34,
      "trailingCommaStart": 57,
      "type": "object-property",
      "value": Object {
        "end": 57,
        "isDynamic": false,
        "parsedLength": 6,
        "quote": "\\"",
        "segments": Array [
          Object {
            "end": 56,
            "parsedLength": 4,
            "raw": "test",
            "start": 52,
            "type": "string-segment",
          },
        ],
        "start": 51,
        "type": "string",
      },
    },
    Object {
      "colonStart": 76,
      "end": 84,
      "hasColon": true,
      "hasTrailingComma": true,
      "key": Object {
        "end": 76,
        "expression": " sym ",
        "expressionEnd": 74,
        "expressionStart": 69,
        "parsedLength": 9,
        "start": 67,
        "type": "expression",
      },
      "parsedLength": 17,
      "start": 67,
      "trailingCommaStart": 83,
      "type": "object-property",
      "value": Object {
        "end": 83,
        "items": Array [
          Object {
            "end": 80,
            "parsedLength": 1,
            "raw": "1",
            "start": 79,
            "type": "number",
            "value": 1,
          },
          Object {
            "end": 82,
            "parsedLength": 1,
            "raw": "2",
            "start": 81,
            "type": "number",
            "value": 2,
          },
        ],
        "itemsLocation": Array [
          Object {
            "hasTrailingComma": true,
            "hasValue": true,
            "trailingCommaStart": 80,
            "valueEnd": 80,
            "valueStart": 79,
          },
          Object {
            "hasValue": true,
            "valueEnd": 82,
            "valueStart": 81,
          },
        ],
        "length": 2,
        "parsedLength": 5,
        "start": 78,
        "type": "array",
      },
    },
    Object {
      "colonStart": 98,
      "end": 104,
      "hasColon": true,
      "hasTrailingComma": false,
      "key": Object {
        "end": 98,
        "isDynamic": false,
        "parsedLength": 5,
        "quote": "\\"",
        "segments": Array [
          Object {
            "end": 97,
            "parsedLength": 3,
            "raw": "bar",
            "start": 94,
            "type": "string-segment",
          },
        ],
        "start": 93,
        "type": "string",
      },
      "parsedLength": 11,
      "start": 93,
      "type": "object-property",
      "value": Object {
        "end": 104,
        "parsedLength": 4,
        "raw": "true",
        "start": 100,
        "type": "boolean",
        "value": true,
      },
    },
  ],
  "start": 7,
  "type": "object",
}
`;

exports[`parse parse object: empty object 1`] = `
Object {
  "end": 2,
  "parsedLength": 2,
  "properties": Array [],
  "start": 0,
  "type": "object",
}
`;

exports[`parse parse static string: empty string 1`] = `
Object {
  "end": 2,
  "isDynamic": false,
  "parsedLength": 2,
  "quote": "\\"",
  "segments": Array [
    Object {
      "end": 1,
      "parsedLength": 0,
      "raw": "",
      "start": 1,
      "type": "string-segment",
    },
  ],
  "start": 0,
  "type": "string",
}
`;

exports[`parse parse static string: empty string 2`] = `
Object {
  "end": 2,
  "isDynamic": false,
  "parsedLength": 2,
  "quote": "'",
  "segments": Array [
    Object {
      "end": 1,
      "parsedLength": 0,
      "raw": "",
      "start": 1,
      "type": "string-segment",
    },
  ],
  "start": 0,
  "type": "string",
}
`;

exports[`parse remainder, {"ignoreUnparsedRemainder":true}: ast 1`] = `
Object {
  "end": 11,
  "expression": " expr ",
  "expressionEnd": 9,
  "expressionStart": 3,
  "parsedLength": 10,
  "start": 1,
  "type": "expression",
}
`;

exports[`parse remainder, {"ignoreUnparsedRemainder":true}: looseModeEnabled 1`] = `false`;

exports[`parse remainder, {"loose":true}: ast 1`] = `
Object {
  "end": 37,
  "isDynamic": true,
  "parsedLength": 36,
  "quote": "",
  "segments": Array [
    Object {
      "end": 1,
      "parsedLength": 0,
      "raw": "",
      "start": 1,
      "type": "string-segment",
    },
    Object {
      "end": 11,
      "expression": " expr ",
      "expressionEnd": 9,
      "expressionStart": 3,
      "parsedLength": 10,
      "start": 1,
      "type": "expression",
    },
    Object {
      "end": 37,
      "parsedLength": 26,
      "raw": ", welcome ! // not-comment",
      "start": 11,
      "type": "string-segment",
    },
  ],
  "start": 1,
  "type": "string",
}
`;

exports[`parse remainder, {"loose":true}: looseModeEnabled 1`] = `true`;
